pkg/runtime/unstructured.go

```diff
@@ -30,13 +33,96 @@ var UnstructuredJSONScheme Codec = unstructuredJSONScheme{}
 
 type unstructuredJSONScheme struct{}
 
-func (s unstructuredJSONScheme) Decode(data []byte, _ *unversioned.GroupVersionKind, _ Object) (Object, *unversioned.GroupVersionKind, error) {
+func (s unstructuredJSONScheme) Decode(data []byte, _ *unversioned.GroupVersionKind, obj Object) (Object, *unversioned.GroupVersionKind, error) {
+	var err error
+	if obj != nil {
+		err = s.decodeInto(data, obj)
+	} else {
+		obj, err = s.decode(data)
+	}
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	if len(gvk.Kind) == 0 {
+		return nil, gvk, NewMissingKindErr(string(data))
+	}
+
+	return obj, gvk, nil
+}
+
+func (unstructuredJSONScheme) EncodeToStream(obj Object, w io.Writer, overrides ...unversioned.GroupVersion) error {
+	switch t := obj.(type) {
+	case *Unstructured:
+		return json.NewEncoder(w).Encode(t.Object)
+	case *UnstructuredList:
+		type encodeList struct {
+			TypeMeta `json:",inline"`
+			Items    []map[string]interface{} `json:"items"`
+		}
+		eList := encodeList{
+			TypeMeta: t.TypeMeta,
+		}
+		for _, i := range t.Items {
+			eList.Items = append(eList.Items, i.Object)
+		}
+		return json.NewEncoder(w).Encode(eList)
+	case *Unknown:
+		_, err := w.Write(t.RawJSON)
+		return err
+	default:
+		return json.NewEncoder(w).Encode(t)
+	}
+}
+
+func (s unstructuredJSONScheme) decode(data []byte) (Object, error) {
+	type detector struct {
+		TypeMeta `json:",inline"`
+		Items    json.RawMessage
+	}
+	var det detector
+	if err := json.Unmarshal(data, &det); err != nil {
+		return nil, err
+	}
+
+	if strings.ToLower(det.Kind) == "status" {
```

Comment:
If anything, you should be doing a second decode of unversioned.Status in the client library in the event you don't receive the data, or at minimum wrap the codec you pass into that type.